zeebe/exporters/elasticsearch-exporter/src/main/java/io/camunda/zeebe/exporter/ElasticsearchMetrics.java

```diff
@@ -7,66 +7,71 @@
 
-  private static final Histogram FLUSH_DURATION =
-      Histogram.build()
-          .namespace(NAMESPACE)
-          .name("flush_duration_seconds")
-          .help("Flush duration of bulk exporters in seconds")
-          .labelNames(PARTITION_LABEL)
-          .register();
+  private final String partitionIdLabel;
+  private final MeterRegistry meterRegistry;
+  private final AtomicInteger bulkMemorySize = new AtomicInteger(0);
+  private final Timer flushDuration;
+  private final DistributionSummary bulkSize;
+  private final Counter failedFlush;
 
-  private static final Counter FAILED_FLUSH =
-      Counter.build()
-          .namespace(NAMESPACE)
-          .name("failed_flush")
-          .help("Number of failed flush operations")
-          .labelNames(PARTITION_LABEL)
-          .register();
+  public ElasticsearchMetrics(final int partitionId, final MeterRegistry registry) {
+    partitionIdLabel = String.valueOf(partitionId);
+    meterRegistry = registry;
 
-  private static final Histogram BULK_SIZE =
-      Histogram.build()
-          .namespace(NAMESPACE)
-          .name("bulk_size")
-          .help("Exporter bulk size")
-          .buckets(10, 100, 1_000, 10_000, 100_000)
-          .labelNames(PARTITION_LABEL)
-          .register();
+    Gauge.builder(meterName("bulk.memory.size"), bulkMemorySize, AtomicInteger::get)
+        .tags(PARTITION_LABEL, partitionIdLabel)
+        .description("Exporter bulk memory size")
+        .register(meterRegistry);
 
-  private static final Gauge BULK_MEMORY_SIZE =
-      Gauge.build()
-          .namespace(NAMESPACE)
-          .name("bulk_memory_size")
-          .help("Exporter bulk memory size")
-          .labelNames(PARTITION_LABEL)
-          .register();
+    flushDuration =
+        Timer.builder(meterName("flush.duration.seconds"))
+            .description("Flush duration of bulk exporters in seconds")
+            .tags(PARTITION_LABEL, partitionIdLabel)
+            .publishPercentileHistogram()
+            .register(meterRegistry);
 
-  private final String partitionIdLabel;
+    bulkSize =
+        DistributionSummary.builder(meterName("bulk.size"))
+            .description("Exporter bulk size")
+            .tags(PARTITION_LABEL, partitionIdLabel)
+            .serviceLevelObjectives(10, 100, 1_000, 10_000, 100_000)
+            .register(meterRegistry);
 
-  public ElasticsearchMetrics(final int partitionId) {
-    partitionIdLabel = String.valueOf(partitionId);
+    failedFlush =
+        Counter.builder(meterName("failed.flush"))
+            .description("Number of failed flush operations")
+            .tags(PARTITION_LABEL, partitionIdLabel)
+            .register(meterRegistry);
   }
 
-  public Histogram.Timer measureFlushDuration() {
-    return FLUSH_DURATION.labels(partitionIdLabel).startTimer();
+  public void measureFlushDuration(final Runnable flushFunction) {
+    flushDuration.record(flushFunction);
```

Comment:
:thought_balloon: I would've liked to see the same pattern as we previously had. From what I see it sort of exists with the Timer.Sample ([see for example here](https://docs.micrometer.io/micrometer/reference/concepts/meter-provider.html)), but then you need to close it with a reference to the timer itself, so unless we build our own wrapper, I don't see this happening =/

I guess there will be a bit of a learning curve on how best to use this.

Reply:
I dislike exposing the implementation details of how the metric is measured if a timer object was returned to the user as it would make it harder to change any backend logic or implementation. Is there a difference I'm not understanding between the old try-with syntax that returns a closable from the method and this method that takes a runnable? Or am I just misunderstanding and you prefer an explicit timer.start -> timer.stop

Reply:
The pattern with an explicit start/stop - or a closeable - is generally just more flexible, especially when coupled with asynchronous code. Here in particular it's only marginally more useful, but it's good to have in general as it's just more flexible.